src/lib/parsers/sol-json.c

```diff
@@ -138,6 +141,193 @@ check_number(struct sol_json_scanner *scanner, struct sol_json_token *token)
+                *value = INT64_MAX;
+            else if (sign < 0 && tmpvar > ((uint64_t)INT64_MAX + 1))
+                *value = INT64_MIN;
+            else
+                *value = sign * tmpvar;
+        }
+        return r;
+    }
+}
+
+SOL_API int
+sol_json_token_get_uint64(const struct sol_json_token *token, uint64_t *value)
+{
+    *value = 0;
+    SOL_NULL_CHECK(token, -EINVAL);
+    SOL_NULL_CHECK(value, -EINVAL);
+    if (token->start >= token->end) {
+        SOL_WRN("invalid token: start=%p, end=%p",
+            token->start, token->end);
+        return -EINVAL;
+    }
+    if (sol_json_token_get_type(token) != SOL_JSON_TYPE_NUMBER) {
+        SOL_WRN("expected number, got token type '%c' for token \"%.*s\"",
+            sol_json_token_get_type(token),
+            sol_json_token_get_size(token), token->start);
+        return -EINVAL;
+    }
+    if (*token->start == '-') {
+        SOL_DBG("%.*s: negative number where unsigned is expected",
+            sol_json_token_get_size(token), token->start);
+        return -ERANGE;
+    }
+
+    return token_get_uint64(token, value);
+}
+
+SOL_API int
+sol_json_token_get_int64(const struct sol_json_token *token, int64_t *value)
+{
+    *value = 0;
+    SOL_NULL_CHECK(token, -EINVAL);
+    SOL_NULL_CHECK(value, -EINVAL);
+    if (token->start >= token->end) {
+        SOL_WRN("invalid token: start=%p, end=%p",
+            token->start, token->end);
+        return -EINVAL;
+    }
+    if (sol_json_token_get_type(token) != SOL_JSON_TYPE_NUMBER) {
+        SOL_WRN("expected number, got token type '%c' for token \"%.*s\"",
+            sol_json_token_get_type(token),
+            sol_json_token_get_size(token), token->start);
+        return -EINVAL;
+    }
+
+    return token_get_int64(token, value);
+}
+
+SOL_API int
+sol_json_token_get_double(const struct sol_json_token *token, double *value)
+{
+    char *endptr, *oldlocale;
+    int r;
+
+    /* NOTE: Using a copy to ensure trailing \0 and strtod() so we
+     * properly parse numbers with large precision.
+     *
+     * Splitting the integer, fractional and exponent parts and doing
+     * the math using double numbers will result in rounding errors
+     * when parsing DBL_MAX using "%.64g" formatting.
+     *
+     * Since parsing it is complex (ie:
+     * http://www.netlib.org/fp/dtoa.c), we take the short path to
+     * call our helper around libc's strtod() that limits the amount
+     * of bytes.
+     *
+     * Extra care is needed since strtod() is locale-dependent, in
+     * pt_BR it will use ',' as fractional/decimal separator and we do
+     * not want that as JSON uses '.' in the same way as the 'C'
+     * locale. Then force 'C' and save oldlocale, restoring it before
+     * we return.
+     */
+    oldlocale = setlocale(LC_NUMERIC, NULL);
+    if (oldlocale && !streq(oldlocale, "C")) {
+        oldlocale = strdupa(oldlocale);
+        setlocale(LC_NUMERIC, "C");
+    }
```

Comment:
The problem in doing this is that this isn't thread safe. We don't have threads now (apart from the worker threads), but we might in the future and this might bite us back. It might be a good idea to use `localeconv()`, get its `decimal_point` character, and if it's not `"."`, substitute it in the passed string before calling `strtod()`.